packages/flutter_migrate/lib/executable.dart

```diff
@@ -0,0 +1,103 @@
+import 'src/base_dependencies.dart';
+
+import 'src/commands/abandon.dart';
+import 'src/commands/apply.dart';
+import 'src/commands/start.dart';
+import 'src/commands/status.dart';
+
+Future<void> main(List<String> args) async {
+  final bool veryVerbose = args.contains('-vv');
+  final bool verbose =
+      args.contains('-v') || args.contains('--verbose') || veryVerbose;
+
+  final MigrateBaseDependencies baseDependencies = MigrateBaseDependencies();
+
+  final List<MigrateCommand> commands = <MigrateCommand>[
+    MigrateStartCommand(
+      verbose: verbose,
+      logger: baseDependencies.logger,
+      fileSystem: baseDependencies.fileSystem,
+      processManager: baseDependencies.processManager,
+    ),
+    MigrateStatusCommand(
+      verbose: verbose,
+      logger: baseDependencies.logger,
+      fileSystem: baseDependencies.fileSystem,
+      processManager: baseDependencies.processManager,
+    ),
+    MigrateAbandonCommand(
+        logger: baseDependencies.logger,
+        fileSystem: baseDependencies.fileSystem,
+        terminal: baseDependencies.terminal,
+        processManager: baseDependencies.processManager),
+    MigrateApplyCommand(
+        verbose: verbose,
+        logger: baseDependencies.logger,
+        fileSystem: baseDependencies.fileSystem,
+        terminal: baseDependencies.terminal,
+        processManager: baseDependencies.processManager),
+  ];
+
+  final MigrateCommandRunner runner = MigrateCommandRunner();
+
+  commands.forEach(runner.addCommand);
+  await runner.run(args);
+
+  await _exit(0, baseDependencies,
+      shutdownHooks: baseDependencies.fileSystem.shutdownHooks);
+  await baseDependencies.fileSystem.dispose();
+}
+
+/// Simple extension of a CommandRunner to provide migrate specific global flags.
+class MigrateCommandRunner extends CommandRunner<void> {
+  MigrateCommandRunner()
+      : super(
+          'flutter',
+          'Migrates legacy flutter projects to modern versions.',
+        ) {
+    argParser.addFlag('verbose',
+        abbr: 'v',
+        negatable: false,
+        help: 'Noisy logging, including all shell commands executed.');
+  }
+
+  @override
+  ArgParser get argParser => _argParser;
+  final ArgParser _argParser = ArgParser();
+}
+
+Future<int> _exit(int code, MigrateBaseDependencies baseDependencies,
+    {required ShutdownHooks shutdownHooks}) async {
+  // Run shutdown hooks before flushing logs
+  await shutdownHooks.runShutdownHooks(baseDependencies.logger);
+
+  final Completer<void> completer = Completer<void>();
+
+  // Give the task / timer queue one cycle through before we hard exit.
+  Timer.run(() {
+    try {
+      baseDependencies.logger.printTrace('exiting with code $code');
+      io.exit(code);
+      // This catches all exceptions because the error is propagated on the
+      // completer.
+    } catch (error, stackTrace) {
+      // ignore: avoid_catches_without_on_clauses
+      completer.completeError(error, stackTrace);
```

Comment:
I always get confused with async error-handling: what's the difference between what you're doing, catching the error and delegating to the completer, and just not handling it? Wouldn't it bubble up and crash the tool regardless?

Reply:
This paradigm is replicated over from what the tools repo does, this specifically comes from `runnable.dart`. Empirically, this specific configuration somehow prevents hangs at the end of program due to open ports, but I have not experimented enough with it to fully see why.

Reply:
lol, i believe you, this is fine

Reply:
although, do you mean without the try-catch the program hangs at the end? are we actually catching something?

Reply:
Let me try it without. I'll report back!

Reply:
Seems like it works without the try. I'll remove it as it's extraneous.

Reply:
SGTM